### PR TITLE
[Improve] pool sendRequest to improve producer perf

### PR DESCRIPTION
**Pull request by gunli**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #<xyz>

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

`sendRequest ` in producer is a frequently allocated struct, pool it can decrease the memory allocation.

### Modifications

1. Init a sync.Pool;
2. Get sendRequest  from the pool when we need; 
3. Reset sendRequest and put it back into the pool when it is done.

Benchmark test:
``` go
package test_test

import (
	"context"
	"github.com/apache/pulsar-client-go/pulsar"
	_ "net/http/pprof"
	"os"
	"runtime/pprof"
	"sync/atomic"
	"testing"
)

var (
	client   pulsar.Client
	producer pulsar.Producer
)

func init() {
	var err error
	client, err = pulsar.NewClient(pulsar.ClientOptions{
		URL: "pulsar://127.0.0.1:6650",
	})
	if err != nil {
		panic(err)
	}
	producer, err = client.CreateProducer(pulsar.ProducerOptions{
		Topic: "tdm_kv",
	})
	if err != nil {
		panic(err)
	}
}

func BenchmarkPulsarSend(b *testing.B) {
	heapFile, err := os.Create("heap.prof")
	if err != nil {
		b.Errorf("error:%v", err)
	}

	payload := []byte("test producer.SendAsync(context.Background(), &pulsar.ProducerMessage{Payload: payload}, go test -bench=. -benchmem callback)producer.SendAsync(context.Background()producer.SendAsync(context.Background()producer.SendAsync(context.Background()producer.SendAsync(context.Background()producer.SendAsync(context.Background(), &pulsar.ProducerMessage{Payload: payload}, callback)")
	send := b.N
	sent := atomic.Int64{}
	done := make(chan struct{})
	callback := func(id pulsar.MessageID, message *pulsar.ProducerMessage, err error) {
		sent.Add(1)
		if sent.Load() == int64(send) {
			done <- struct{}{}
		}
	}

	for i := 0; i < b.N; i++ {
		producer.SendAsync(context.Background(), &pulsar.ProducerMessage{Payload: payload}, callback)
	}

	<-done
	if err := pprof.WriteHeapProfile(heapFile); err != nil {
		b.Errorf("error:%v", err)
	}
	_ = heapFile.Close()
}

```
Benchmark with this PR:
``` shell
go test -bench=. -benchmem -benchtime=20s
goos: linux
goarch: amd64
pkg: pulsar-test/test
cpu: Intel(R) Xeon(R) Gold 6133 CPU @ 2.50GHz
BenchmarkPulsarSend-16           7712277              4678 ns/op            1710 B/op         12 allocs/op
PASS
ok      pulsar-test/test        39.294s

```

Benchmark without this PR:
``` shell
go test -bench=. -benchmem -benchtime=20s
goos: linux
goarch: amd64
pkg: pulsar-test/test
cpu: Intel(R) Xeon(R) Gold 6133 CPU @ 2.50GHz
BenchmarkPulsarSend-16           7769998              4724 ns/op            2068 B/op         13 allocs/op
PASS
ok      pulsar-test/test        39.866s

```

We can see that when enabling this PR, we can decrease 1 allocs/op and about 350 B/op.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)

### Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / GoDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by gunli** [comment]
@RobertIndie @tisonkun @Gleiphir2769 @shibd

**Comment by Gleiphir2769** [comment]
Hi @gunli, the implement looks good to me. But most member of `SendRequest` are pointer. I think It's not essential to pool it. Could you give a more significant benchmark result?

**Comment by gunli** [comment]
> Hi @gunli, the implement looks good to me. But most member of `SendRequest` are pointer. I think It's not essential to pool it. Could you give a more significant benchmark result?

heap-pr
heap-nopr

Yes, here are the heap pprof results when benchmarked with and without this PR, we can see when benchmarked  without this PR, `internalSend` and `internalSendAsync`, in where `sendRequest`s are allocated, allocated much more memory than that when it benchmarked with this PR.

**Comment by Gleiphir2769** [comment]
>  we can see when benchmarked without this PR, internalSend and internalSendAsync, in where sendRequests are allocated, allocated much more memory than that when it benchmarked with this PR.
Good work @gunli ! I have no doubt now.

**Comment by gunli** [comment]
For the sake of fairness, I updated the benchmark, set b.N = 500000, the pprof reports are:

pr
no-pr

File: pulsar/producer_partition.go [hunk 1]
```diff
@@ -1477,6 +1491,13 @@ func (sr *sendRequest) done(msgID MessageID, err error) {
 			sr.transaction.endSendOrAckOp(err)
 		}
 	}
+
+	pool := sr.pool
+	if pool != nil {
+		// reset all the fields
+		*sr = sendRequest{}
```
Comment (rfyiamcool) [inline pulsar/producer_partition.go:1498]: 👍

**Review by tisonkun — APPROVED** [review]
Thank you!

cc @RobertIndie may you take a look also for preventing regression.

**Review by RobertIndie — APPROVED** [review]
LGTM!

**Pull request merged**